File: Classes/Controls/DocumentView/Text/DocumentTextView.m
//
//  DocumentTextView.m
//  Dokie
//
//  Created by Ashamaz Shidov on 21/12/2019.
//  Copyright © 2019 Ashamaz Shidov. All rights reserved.
//

#import "DocumentTextView.h"
#import "DocumentTextViewModel.h"
#import "Definitions.h"
#import "DocumentTextTableViewCell.h"
#import "UI.h"
#import "AppearanceManager.h"

@interface DocumentTextView() <UITableViewDelegate, UITableViewDataSource>

@property (nonatomic, strong) DocumentTextViewModel *viewModel;
@property (nonatomic, strong) NSMutableArray<NSIndexPath *> *selectedCells;
@property (nonatomic) NSInteger selectedCount;

@end

@implementation DocumentTextView

- (instancetype)initWithViewModel:(DocumentTextViewModel *)viewModel {
    self = [super init];
    if (self) {
        self.viewModel = viewModel;
        [self setup];
    }
    return self;
}

- (void)setup {
    @weakify(self);
    self.selectedCells = [NSMutableArray array];
    self.selectedCount = self.selectedCells.count;
    self.backgroundColor = [AppearanceManager backgroundColor];
    
    UITableView *tableView = [UI tableView];
    tableView.allowsMultipleSelectionDuringEditing = NO;
    tableView.delegate = self;
    tableView.dataSource = self;
    [self addSubview:tableView];
    
    NSDictionary *views = NSDictionaryOfVariableBindings(tableView);
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"V:|[tableView]|" options:0 metrics:nil views:views]];
    [self addConstraints:[NSLayoutConstraint constraintsWithVisualFormat:@"H:|[tableView]|" options:0 metrics:nil views:views]];
    
    UILongPressGestureRecognizer *longPress = [[UILongPressGestureRecognizer alloc] initWithTarget:tableView action:nil];
    longPress.minimumPressDuration = 1.0;
    @weakify(tableView);
    [longPress.rac_gestureSignal subscribeNext:^(__kindof UIGestureRecognizer *gestureRecognizer) {
        @strongify(self, tableView);

        if (gestureRecognizer.state == UIGestureRecognizerStateBegan) {
            CGPoint p = [gestureRecognizer locationInView:tableView];
            NSIndexPath *indexPath = [tableView indexPathForRowAtPoint:p];
            [self.viewModel editContentAtIndexPath:indexPath];
        }
    }];
    [tableView addGestureRecognizer:longPress];
    
    [[RACObserve(self, viewModel.content) deliverOnMainThread] subscribeNext:^(id _) {
        @strongify(tableView);
        [tableView reloadData];
    }];
    
    [[[RACObserve(self, viewMode) distinctUntilChanged] deliverOnMainThread] subscribeNext:^(NSNumber *viewMode) {
        @strongify(self, tableView);
        [self.selectedCells removeAllObjects];
        self.selectedCount = self.selectedCells.count;
        if (viewMode.integerValue == ViewModeDisplay) {
            [self.selectedCells removeAllObjects];
            self.selectedCount = self.selectedCells.count;
            for (DocumentTextTableViewCell *cell in tableView.visibleCells) {
                cell.viewMode = self.viewMode;
                cell.showCheckmark = NO;
            }
        }
        else {
            [self.selectedCells removeAllObjects];
            self.selectedCount = self.selectedCells.count;
            for (DocumentTextTableViewCell *cell in tableView.visibleCells) {
                cell.viewMode = self.viewMode;
            }
        }
    }];
}

- (void)refresh {
    [[self.viewModel readDocument] subscribeCompleted:^{}];
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.viewModel.content.count;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    return 50;
}

static NSString *const contentCellIdentifier = @"contentCellIdentifier";

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    DocumentTextTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:contentCellIdentifier];
    if (!cell) {
        cell = [[DocumentTextTableViewCell alloc] initWithStyle:UITableViewCellStyleSubtitle reuseIdentifier:contentCellIdentifier];
        [cell setSelectionStyle:UITableViewCellSelectionStyleNone];
    }
    RACTupleUnpack(NSString *title, NSString *detail) = self.viewModel.content[indexPath.row];
    cell.text = title;
    cell.detail = detail;
    cell.viewMode = self.viewMode;
    cell.showCheckmark = self.viewMode == ViewModeSelect && [self.selectedCells containsObject:indexPath];
    @weakify(self);
    cell.editCommand = [[RACCommand alloc] initWithSignalBlock:^RACSignal * _Nonnull(id _) {
        @strongify(self);
        [self.viewModel editContentAtIndexPath:indexPath];
        return [RACSignal empty];
    }];
    return cell;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    if (self.viewMode == ViewModeDisplay) {
        [self.viewModel didSelectTextAtIndexPath:indexPath];
    }
    else {
        if ([self.selectedCells containsObject:indexPath]) {
            [self.selectedCells removeObject:indexPath];
        }
        else {
            [self.selectedCells addObject:indexPath];
        }
        for (DocumentTextTableViewCell *cell in tableView.visibleCells) {
            if ([[tableView indexPathForCell:cell] isEqual:indexPath]) {
                cell.showCheckmark = [self.selectedCells containsObject:indexPath];
            }
        }
        self.selectedCount = self.selectedCells.count;
    }
}

- (NSArray<NSString *> *)selected {
    NSMutableArray *toReturn = [NSMutableArray array];
    for (NSIndexPath *indexPath in self.selectedCells) {
        RACTupleUnpack(NSString *title, NSString *detail) = self.viewModel.content[indexPath.row];
        if (title.length > 0 && detail.length > 0) {
            [toReturn addObject:[NSString stringWithFormat:@"%@: %@", title, detail]];
        }
        else if (title.length > 0) {
            [toReturn addObject:title];
        }
        else {
            [toReturn addObject:detail];
        }
    }
    return [toReturn copy];
}

- (void)deleteSelected {
    [self.viewModel deleteContentAtIndexPaths:self.selectedCells];
}

@end

File: Classes/Views/DocumentView/Text/DocumentTextView.swift
//
//  DocumentTextView.swift
//  Dokie
//
//  Created by Ashamaz Shidov on 17.05.2020.
//  Copyright © 2020 Ashamaz Shidov. All rights reserved.
//

import UIKit
import ReactiveCocoa
import ReactiveSwift

class DocumentTextView : UIView {
    
    let viewMode = MutableProperty<ViewMode>(.display)
    let selectedCount = MutableProperty<Int>(0)
    
    private let viewModel: DocumentTextViewModel!
    private var selectedCells = [IndexPath]()
    
    init(viewModel: DocumentTextViewModel) {
        self.viewModel = viewModel
        super.init(frame: CGRect())
        setup()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setup() {
        backgroundColor = Appearance.backgroundColor
        
        let tableView = UI.tableView()
        tableView.allowsMultipleSelectionDuringEditing = false
        tableView.delegate = self
        tableView.dataSource = self
        addSubview(tableView)
        
        let views = ["tableView" : tableView]
        addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:|[tableView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: [:], views: views))
        addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[tableView]|", options: NSLayoutConstraint.FormatOptions(rawValue: 0), metrics: [:], views: views))
        
        let longPress = UILongPressGestureRecognizer(target: tableView, action: nil)
        longPress
            .reactive
            .stateChanged
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self, tableView] gesture in
                if gesture.state == .recognized {
                    let point = gesture.location(in: tableView)
                    if let indexPath = tableView.indexPathForRow(at: point) {
                        self.viewModel.edit(indexPath: indexPath)
                    }
                }
        }
        tableView.addGestureRecognizer(longPress)
        
        viewModel
            .content
            .producer
            .skip(first: 1)
            .skipRepeats({ (left, right) -> Bool in
                left == right
            })
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned tableView] _ in
                tableView.reloadData()
        }
        
        viewMode
            .producer
            .take(during: reactive.lifetime)
            .observe(on: UIScheduler())
            .startWithValues { [unowned self, tableView] viewMode in
                self.selectedCells.removeAll()
                self.selectedCount.value = 0
                
                switch viewMode {
                case .display:
                    for cell in tableView.visibleCells {
                        if let tableCell = cell as? DocumentTextTableViewCell {
                            tableCell.viewMode.value = viewMode
                            tableCell.showCheckmark.value = false
                        }
                    }
                case .select:
                    for cell in tableView.visibleCells {
                        if let tableCell = cell as? DocumentTextTableViewCell {
                            tableCell.viewMode.value = viewMode
                        }
                    }
                    
                }
        }
    }
    
    func refresh() {
        viewModel.read().start { _ in }
    }
    
    var selected: [String] {
        var toReturn = [String]()
        for indexPath in selectedCells {
            let content = viewModel.content.value[indexPath.row]
            toReturn.append(content.string())
        }
        
        return toReturn
    }
    
    func deleteSelected() {
        viewModel.delete(indexPaths: selectedCells)
    }
}

extension DocumentTextView : UITableViewDelegate, UITableViewDataSource {
    func numberOfSections(in tableView: UITableView) -> Int {
        1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        viewModel.content.value.count
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        50
    }
    
    private static let contentCellIdentifier = "contentCellIdentifier"
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: DocumentTextTableViewCell
        if let tableCell = tableView.dequeueReusableCell(withIdentifier: DocumentTextView.contentCellIdentifier) as? DocumentTextTableViewCell {
            cell = tableCell
        }
        else {
            cell = DocumentTextTableViewCell(style: .subtitle, reuseIdentifier: DocumentTextView.contentCellIdentifier)
            cell.selectionStyle = .none
        }
        
        let value = viewModel.content.value[indexPath.row]
        cell.textProperty.value = value.text
        cell.descriptionProperty.value = value.description
        cell.viewMode.value = viewMode.value
        cell.showCheckmark.value = viewMode.value == .select && selectedCells.contains(indexPath)
        let editAction = Action<(), UIButton, Never> { _ -> SignalProducer<UIButton, Never> in
            return SignalProducer { [unowned self] (observer, disposable) in
                self.viewModel.edit(indexPath: indexPath)
                observer.sendCompleted()
            }
        }
        
        cell.editCommand = CocoaAction(editAction)
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        switch viewMode.value {
        case .display:
            viewModel.didSelect(indexPath: indexPath)
        case .select:
            let showCheckmark: Bool
            if let index = selectedCells.firstIndex(of: indexPath) {
                selectedCells.remove(at: index)
                showCheckmark = false
            }
            else {
                selectedCells.append(indexPath)
                showCheckmark = true
            }
            
            for cell in tableView.visibleCells {
                if let tableCell = cell as? DocumentTextTableViewCell,
                    tableView.indexPath(for: cell) == indexPath {
                    tableCell.showCheckmark.value = showCheckmark
                }
            }
            selectedCount.value = selectedCells.count
        }
    }
    
}
